File: corporate/script/seed.mjs
import {readFileSync} from 'node:fs';
import {dirname, join} from 'node:path';
import {fileURLToPath} from 'node:url';
import {load} from './load.mjs';

const corporate = dirname(dirname(fileURLToPath(import.meta.url)));
const fixtures = join(corporate, 'fixtures');
const port = Number(process.env.CORP_PORT || 8790);
const g = load({drive: process.env.CORP_DATA});

function plant(name) {
  const fixture = JSON.parse(readFileSync(join(fixtures, name), 'utf8'));
  const questions = g.load_questions().questions;
  const event = g.create_event(fixture.name, fixture.company, fixture.date, fixture.tz, fixture.people);
  const eventId = event.event_id;
  const answers = Object.fromEntries(Object.entries(fixture.answers).map(([id, value]) => [id, {value, state: 'confirmed'}]));
  for (const state of ['unknown', 'miles', 'none']) {
    for (const id of fixture[state] || []) answers[id] = {value: null, state};
  }
  let result = g.save(eventId, 'p_miles', 'dj', {
    base_revision: event.revision,
    submission_id: `sub_seed_${eventId}`,
    answers,
  }, questions);
  if (result[0] !== 200) throw new Error(`the fixture would not save: ${JSON.stringify(result[1])}`);
  result = g.seed_apply(eventId, fixture.moments, fixture.proposals || [], fixture.dj_notes, fixture.stage, 'p_miles', questions);
  if (result[0] !== 200) throw new Error(`the moments would not save: ${JSON.stringify(result[1])}`);
  const links = fixture.people.map((person) => {
    const grant = g.access('mint', {event_id: eventId, person_id: person.person_id, role: person.role});
    return [person.name, person.role, grant.token];
  });
  return [g.load_event(eventId), links];
}

g.wipe();
console.log('Store emptied.\n');
for (const name of ['harbor-studio.json', 'northstar-awards.json']) {
  const [event, links] = plant(name);
  console.log(event.answers.event_name.value);
  console.log(`  ${event.next_action}`);
  console.log(`  ${event.moments.filter((moment) => moment.active !== false).length} parts of the night, ${event.open_items.filter((item) => item.active !== false && !item.resolved).length} questions still open.`);
  for (const [who, role, token] of links) {
    console.log(`  ${who.padEnd(16)} ${role.padEnd(11)} http://127.0.0.1:${port}/corporate/client/#${token}`);
  }
  console.log();
}
const dj = g.access('dj', {});
console.log(`Miles's own view:  http://127.0.0.1:${port}/corporate/dj/`);
console.log(`His pass, for the page to carry:  ${dj.token}`);
console.log('\nStart the server with:  node corporate/script/standin.mjs');

File: corporate/seed.py
"""Empty the store and put the two pretend events back in it.

Run this whenever you want a clean pair of events to look at.  It prints one
private link per person and the link for Miles's own view.  Everybody in here
is invented; none of it is a real client.
"""

import json
import sys
from pathlib import Path

import store

HERE = Path(__file__).resolve().parent
FIXTURES = HERE / "fixtures"


def load_fixture(name):
    return json.loads((FIXTURES / name).read_text(encoding="utf-8"))


def plant(fixture):
    questions = store.load_questions()["questions"]
    event = store.create_event(fixture["name"], fixture["company"],
                               fixture["date"], fixture["tz"],
                               people=fixture["people"])
    event_id = event["event_id"]

    answers = {qid: {"value": value, "state": "confirmed"}
               for qid, value in fixture["answers"].items()}
    for state in ("unknown", "miles", "none"):
        for qid in fixture.get(state, []):
            answers[qid] = {"value": None, "state": state}
    code, body = store.save(event_id, "p_miles", "dj",
                            {"base_revision": event["revision"],
                             "submission_id": "sub_seed_" + event_id,
                             "answers": answers}, questions)
    if code != 200:
        raise SystemExit("the fixture would not save: %s" % json.dumps(body, indent=2))

    code, body = store.seed_apply(event_id, moments=fixture["moments"],
                                  proposals=fixture.get("proposals", []),
                                  dj_notes=fixture.get("dj_notes"),
                                  stage=fixture.get("stage"), questions=questions)
    if code != 200:
        raise SystemExit("the moments would not save: %s" % json.dumps(body, indent=2))

    links = []
    for person in fixture["people"]:
        grant = store.access("mint", event_id=event_id,
                             person_id=person["person_id"], role=person["role"])
        links.append((person["name"], person["role"], grant["link"]))
    return store.load_event(event_id), links


def main():
    store.wipe()
    print("Store emptied.\n")
    for name in ("harbor-studio.json", "northstar-awards.json"):
        fixture = load_fixture(name)
        event, links = plant(fixture)
        print(event["answers"]["event_name"]["value"])
        print("  " + event["next_action"])
        print("  %d parts of the night, %d questions still open."
              % (len([m for m in event["moments"] if m.get("active", True)]),
                 len([i for i in event["open_items"]
                      if i.get("active", True) and not i.get("resolved")])))
        for who, role, link in links:
            print("  %-16s %-11s http://127.0.0.1:8790%s" % (who, role, link))
        print()
    dj = store.access("dj")
    print("Miles's own view:  http://127.0.0.1:8790/dj/")
    print("His pass, for the page to carry:  %s" % dj["token"])
    print("\nStart the server with:  python3 corporate/server.py")
    return 0


if __name__ == "__main__":
    sys.exit(main())
